### PR TITLE
Enable the assert NAN tests and get them passing

### LINKED ISSUE #94: Enable assertarithmaticnan spec tests
https://github.com/ethereum/py-wasm/blob/6559ef319126b4874a03ba9468e9d1eb7e066022/wasm/tools/fixtures/runner.py#L305-L318

Just realized there's a set of spec tests that are still not implemented.

### PR DESCRIPTION
**Pull request by pipermerriam**
fixes #94 

## What was wrong?

There were still some spec tests being skiped, specifically the ones that do assertions about returned NAN values.

## How was it fixed?

Enabled the tests and fixed the logic function implementations so that they passed.

#### Cute Animal Picture

animal-animals-babies-bird-birds-cute-favim com-40474

### CODE REVIEW

File: wasm/_utils/float.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+)
+
+import numpy
+
+from wasm import (
+    constants,
+)
+from wasm.typing import (
+    AnyFloat,
+    AnyUnsigned,
```
Comment (carver) [inline wasm/_utils/float.py]: Maybe `UnsignedFloat`? `Any` name still makes me think it includes integers.

File: wasm/_utils/float.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        (int(sign) << 63) | (int(exponent) << EXPONENT_64_SHIFT) | int(mantissa)
+    )
+    return numpy.frombuffer(as_uint64.data, numpy.float64)[0]
+
+
+def compose_float32(sign: AnyUnsigned,
+                    exponent: AnyUnsigned,
+                    mantissa: AnyUnsigned) -> numpy.float32:
+    as_uint32 = numpy.uint32(
+        (int(sign) << 32) | (int(exponent) << EXPONENT_32_SHIFT) | int(mantissa)
```
Comment (carver) [inline wasm/_utils/float.py]: Should this be a 31 bit shift? If so, maybe there's a test case missing.

File: wasm/_utils/float.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        return _decompose_float32(value)
+    else:
+        raise TypeError(f"Invalid type: {type(value)}")
+
+
+def compose_float64(sign: AnyUnsigned,
+                    exponent: AnyUnsigned,
+                    mantissa: AnyUnsigned) -> numpy.float64:
+    as_uint64 = numpy.uint64(
+        (int(sign) << 63) | (int(exponent) << EXPONENT_64_SHIFT) | int(mantissa)
```
Comment (carver) [inline wasm/_utils/float.py]: Would be nice to move the sign shift magic number into a constant, both here and in float32.

File: wasm/logic/numeric.py [hunk 4]
```diff
@@ -710,7 +710,8 @@ def fceil_op(config: Configuration) -> None:
     logger.debug("%s(%s)", instruction.opcode.text, value)
 
     if numpy.isnan(value):
-        config.push_operand(value)
+        with allow_invalid():
+            config.push_operand(numpy.ceil(value))
```
Comment (carver) [inline wasm/logic/numeric.py]: Huh, interesting, does `ceil` actually do something here? Or is this equivalent to `push_operand(F64_CANON_NAN)`?

Just not sure what allowing invalid means in this context. If it's a no-op then it seems like it's not worth even running the `numpy.ceil()` function.

File: wasm/logic/numeric.py [hunk 5]
```diff
@@ -1137,7 +1152,15 @@ def f64promote_op(config: Configuration) -> None:
 
     logger.debug("%s(%s)", instruction.opcode.text, value)
 
-    config.push_operand(numpy.float64(value))
+    if numpy.isnan(value):
+        if _is_negative(value):
+            logger.info('NEGATIVE')
+            config.push_operand(numpy.float64('-nan'))
+        else:
+            logger.info('POSITIVE')
```
Comment (carver) [inline wasm/logic/numeric.py]: Looks like some cruft to be removed.

File: wasm/tools/fixtures/runner.py [hunk 6]
```diff
@@ -172,14 +176,18 @@ def run_opcode_action_get(action: Action,
...
+
+
+def do_assert_return(command: TDoAssertReturnCommands,
                      module: CurrentModule,
                      all_modules: AllModules,
                      runtime: Runtime) -> None:
-    try:
-        ret = run_opcode_action(command, module, all_modules, runtime)
-    except FloatingPointNotImplemented:
-        return
```
Comment (carver) [inline wasm/tools/fixtures/runner.py]: :tada:

File: wasm/_utils/float.py (hunk 2 above)
Comment (davesque) [reply wasm/_utils/float.py]: Yeah, it should be.

File: wasm/_utils/float.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    AnyFloat,
+    AnyUnsigned,
+)
+
+# the mantissa bits
+MANTISSA_64_MASK = numpy.uint64(2**constants.F64_SIGNIF - 1)
+MANTISSA_32_MASK = numpy.uint32(2**constants.F32_SIGNIF - 1)
+
+# the expon bits
+EXPONENT_64_SHIFT = 64 - 1 - constants.F64_EXPON
```
Comment (davesque) [inline wasm/_utils/float.py]: Alternatively, you could just use `constants.FXX_SIGNIF` for these values.

File: wasm/_utils/float.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    else:
+        raise TypeError(f"Invalid type: {type(value)}")
+
+
+# runtime check since the actual value *might* differ on different platforms.
+assert is_canonical_nan(constants.F32_CANON_NAN)
+assert is_canonical_nan(constants.F64_CANON_NAN)
+
+
+def is_arithmetic_nan(value: Any) -> bool:
```
Comment (davesque) [inline wasm/_utils/float.py]: How is this function used?  It appears to leave out a range of nan values whose mantissas are less than `FXX_CANON_N` but greater than zero.

File: wasm/_utils/float.py (hunk 2 above)
Comment (carver) [reply wasm/_utils/float.py]: Maybe a quick hypothesis test to make sure that `compose(decompose(afloat)) == afloat`

File: wasm/_utils/float.py (hunk 8 above)
Comment (pipermerriam) [reply wasm/_utils/float.py]: It's from this section:

https://webassembly.github.io/spec/core/bikeshed/index.html#floating-point%E2%91%A0

> An arithmetic `NaN` is a floating-point value `±nan(n)` with `n≥canonN`, such that the most significant bit is `1` while all others are arbitrary.

File: wasm/logic/numeric.py (hunk 4 above)
Comment (pipermerriam) [reply wasm/logic/numeric.py]: The `invalid` comes from feeding a `NaN` into `numpy.ceil()`.  Internally, I don't know what the call does, but the tests fail without it.

**Pull request merged**